### ISSUE TITLE
Whether isolation of exit queue with token address (ERC contract address) is good enough

### ISSUE THREAD

**Issue by boolafish** [omisego/plasma-contracts#395]
## Discussion
Currently we isolate different ERC contracts in different exit queues so that any "bad" implementation of the ERC contract would only fail its own queue. However, after #385 shows that even the isolation of ETH queue, there is still much need to be aware on the implementation. So I start to doubt that whether an isolation of queue is good enough. Could there be a "good" implementation but still possibly to fail a `transfer` and block our queue.

Some attacking point that I could think of:

In ERC20 implementation of openzepplin, it [blocks transfer from and to zero address](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L155). I am currently not sure our system accepts transaction to zero address (as a burn) or not, but this is already sth that is depending on the implementation outside our control and should be consider as a "good" implementation. So if anyone exits to a zero address, it would successfully block our queue. Also, if some ERC20 implementation just decide to block `address(1)` what should we do? It is logically not a "bad" contract I would say. 

[edit: ah...it is impossible to exit to burn addresses as we check `msg.sender` equals output owner. However this means if one day we are adding the feature of exit to another address, this would be a blocker]

And if one day we are supporting ERC777, see the send function: [spec here](https://docs.openzeppelin.com/contracts/2.x/api/token/erc777#IERC777-send-address-uint256-bytes-). If the receiver is a contract, it would revert if that contract does not implement some special function. Which means the attack can just be making the output owner a contract without right implementation. Again, it seems to be designed to fail for ERC777 contract but a single failure (even intended) would fail our all queue.

## Suggestion
Reconsider pull-over-push mechanism, which seems to solve #385 too

**Comment by pdobacz** [comment]
Nice! Some thoughts on this:

>  I am currently not sure our system accepts transaction to zero address (as a burn) or not, but this is already sth that is depending on the implementation outside our control and should be consider as a "good" implementation. So if anyone exits to a zero address, it would successfully block our queue. 

I think it accepts such a transaction, but the output is unexitable, since zero address can't call `startStandardExit` nor `piggybackXYZ`, thereby disabling the possibility of a `processExit` on that later on (I think...).

But this is a great point to keep in mind!

> Also, if some ERC20 implementation just decide to block address(1) what should we do? It is logically not a "bad" contract I would say.

There was a discussion (and a ticket to investigate - on pivotal tracker :( ) on this a long time ago. The decision was, that for now any ERC20 restricting transfer is treated as "broken".

So logically it is not "bad" just as you say, but we treat it as "bad" (use at your own risk).

> (ERC777) If the receiver is a contract, it would revert if that contract does not implement some special function

but then I think this would only mean a zero-return from the call to the ERC777 (or otherwise, we still can catch that revert). So this could be up to the `ERC777Vault` implementation.

**Comment by thec00n** [comment]
From the discussion on slack:

> TL;DR, ERC20 of openzepplin implementation blocks address(0), so it would actually block all txs in queue if we accepted addresss(0) and the ERC20 is using openzepplin version of implementation. (edited) 

@boolafish @kevsul @pdobacz can we move ahead and disallow that the  `outputGuard` has a zero value because in current implementation this would break the exit queue?

**Comment by pdobacz** [comment]
> @boolafish @kevsul @pdobacz can we move ahead and disallow that the `outputGuard` has a zero value because in the current implementation this would break the exit queue?

Totally, just the usual suggestion to make it super evident in the docs and tests where does this requirement come from.

**Comment by thec00n** [comment]
@boolafish can we close this issue ?

**Comment by boolafish** [comment]
output guard not being address(0) is handled in PR #502.
For now, there is not really any action items we can take on this. closing this. Just need to be aware whenever we introduce new vault that ERC implementation does impact our assumption.
